### PR TITLE
feat(syslogexporter): add option to skip certificate verfication

### PR DESCRIPTION
**Pull request by kkujawa-sumo**
This allows using self-signed certificates

### CODE REVIEW

**Comment by sumo-drosiek** [comment]
Isn't there any helpers for http client in Otel?

File: pkg/exporter/syslogexporter/README.md [hunk 1]
```diff
@@ -26,7 +24,10 @@ __Note__: Syslog over UDP doesn't support certificate verification (`ca_certific
 - `format` - rfc5424/rfc3164
   - `rfc5424` - Expects the syslog messages to be rfc5424 compliant
   - `rfc3164` - Expects the syslog messages to be rfc3164 compliant
-- `ca_certificate` [tcp only] - A publicly verifiable server certificate (`note`: Self signed certificates are not supported in this version)
+- `ca_certificate` [tcp only] - A publicly verifiable server certificate
+- `certificate` [tcp only] - certificate for mTLS communication (client certificate)
+- `key` [tcp only] -  Key for mTLS communication (client key)
+- `skip_certificate_verification` - [tcp only] If set to `true` then any certificate presented by the server is accepted and any host name in that certificate
```
Comment (sumo-drosiek) [inline pkg/exporter/syslogexporter/README.md:30]: Can we group configuration under common key?

**Comment by kkujawa-sumo** [comment]
> Isn't there any helpers for http client in Otel?

We don't use http in syslogexporter, we use only TCP/UDP.

File: pkg/exporter/syslogexporter/README.md (hunk 1 above)
Comment (kkujawa-sumo) [reply pkg/exporter/syslogexporter/README.md]: I will try to use https://github.com/open-telemetry/opentelemetry-collector/blob/7cf999e4526c9e65f5dbedb67c90cbe2c79e4f54/config/configtls/configtls.go#L66 and make settings similar to configuration described here: https://github.com/open-telemetry/opentelemetry-collector/blob/exporter/otlphttpexporter/v0.71.0/config/configtls/README.md

**Comment by swiatekm-sumo** [comment]
> > Isn't there any helpers for http client in Otel?
> 
> We don't use http in syslogexporter, we use only TCP/UDP.

Otel core does have a helper for TLS though: https://github.com/open-telemetry/opentelemetry-collector/tree/main/config/configtls, which we should use, though not necessarily in this PR.

**Comment by kkujawa-sumo** [comment]
> > > Isn't there any helpers for http client in Otel?
> > 
> > 
> > We don't use http in syslogexporter, we use only TCP/UDP.
> 
> Otel core does have a helper for TLS though: https://github.com/open-telemetry/opentelemetry-collector/tree/main/config/configtls, which we should use, though not necessarily in this PR.

@swiatekm-sumo I mentioned it in this comment:  https://github.com/SumoLogic/sumologic-otel-collector/pull/983#discussion_r1108199994

**Comment by kkujawa-sumo** [comment]
Configuration for TLS is changed in this PR and uses  https://github.com/open-telemetry/opentelemetry-collector/tree/main/config/configtls, I assume that this kind of changes are not treated as breaking changes because syslog exporter has alpha stability https://github.com/open-telemetry/opentelemetry-collector#alpha

**Review by swiatekm-sumo — APPROVED** [review]

File: pkg/exporter/syslogexporter/examples/config_with_token.yaml [hunk 2]
```diff
@@ -8,9 +8,10 @@ exporters:
     protocol: tcp
     port: 514
     endpoint: 127.0.0.1
-    ca_certificate: certs/servercert.pem
-    certificate: certs/cert.pem # tcp only
-    key: certs/key.pem # tcp only
+    tls:
+      ca_file: certs/servercert.pem
```
Comment (rnishtala-sumo) [inline pkg/exporter/syslogexporter/examples/config_with_token.yaml:12]: should we change this name to ca_server_cert?

File: pkg/exporter/syslogexporter/exporter.go [hunk 3]
```diff
@@ -43,27 +42,9 @@ func initExporter(cfg *Config, createSettings exporter.CreateSettings) (*sysloge
...
-	}
-
-	if cfg.Certificate != "" && cfg.Key != "" {
-		var certificate tls.Certificate
-		certificate, err = tls.LoadX509KeyPair(cfg.Certificate, cfg.Key)
-		if err != nil {
-			return nil, err
-		}
-		tlsConfig.Certificates = []tls.Certificate{certificate}
+	tlsConfig, err = cfg.TLSSetting.LoadTLSConfig()
```
Comment (rnishtala-sumo) [inline pkg/exporter/syslogexporter/exporter.go:45]: will this make adding all cert options mandatory?

File: pkg/exporter/syslogexporter/README.md [hunk 4]
```diff
@@ -14,21 +14,25 @@ The syslog exporter supports sending messages to a remote syslog server.
...
   - `rfc3164` - Expects the syslog messages to be rfc3164 compliant
-- `ca_certificate` [tcp only] - A publicly verifiable server certificate (`note`: Self signed certificates are not supported in this version)
-- `certificate` [tcp only] - certificate for mTLS communication (client certificate)
-- `key` [tcp only] -  Key for mTLS communication (client key)
+- `tls` - configuration for TLS/mTLS
+  - `insecure` (default = `false`) whether to enable client transport security, by default, TLS is enabled.
+  - `cert_file` - Path to the TLS cert to use for TLS required connections. Should only be used if `insecure` is set to `false`.
+  - `key_file` - Path to the TLS key to use for TLS required connections. Should only be used if `insecure` is set to `false`.
+  - `ca_file` - Path to the CA cert. For a client this verifies the server certificate. For a server this verifies client certificates. If empty uses system root CA. Should only be used if `insecure` is set to `false`.
+  - `insecure_skip_verify` -  (default = `false`) whether to skip verifying the certificate or not.
```
Comment (rnishtala-sumo) [inline pkg/exporter/syslogexporter/README.md:32]: are these options built into the `configtls`?

File: pkg/exporter/syslogexporter/examples/config_with_token.yaml (hunk 2 above)
Comment (rnishtala-sumo) [reply pkg/exporter/syslogexporter/examples/config_with_token.yaml:12]: actually never mind this I see that t needs to be called `ca_file`

**Review by rnishtala-sumo — APPROVED** [review]

File: pkg/exporter/syslogexporter/exporter.go (hunk 3 above)
Comment (kkujawa-sumo) [reply pkg/exporter/syslogexporter/exporter.go:45]: No, it is as earlier, it creates tls config from configuration https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/configtls/configtls.go

File: pkg/exporter/syslogexporter/README.md (hunk 4 above)
Comment (kkujawa-sumo) [reply pkg/exporter/syslogexporter/README.md:32]: please see here: https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/configtls/configtls.go, these options are used to created tls config

**Pull request merged**